implementations/rust/ockam/ockam_identity/src/secure_channel/decryptor.rs

```diff
@@ -1,13 +1,122 @@
+impl DecryptorHandler {
+    pub fn new(
+        role: &'static str,
+        addresses: Addresses,
+        key: KeyId,
+        vault: Arc<dyn XXInitializedVault>,
+        their_identity_id: IdentityIdentifier,
+    ) -> Self {
+        Self {
+            role,
+            addresses,
+            their_identity_id,
+            decryptor: Decryptor::new(key, vault),
+        }
+    }
+
+    pub(crate) async fn handle_decrypt_api(
+        &mut self,
+        ctx: &mut Context,
+        msg: Routed<Any>,
+    ) -> Result<()> {
+        debug!(
+            "SecureChannel {} received Decrypt API {}",
+            self.role, &self.addresses.decryptor_remote
+        );
+
+        let return_route = msg.return_route();
+
+        // Decode raw payload binary
+        let request = DecryptionRequest::decode(&msg.into_transport_message().payload)?;
+
+        // Decrypt the binary
+        let decrypted_payload = self.decryptor.decrypt(&request.0).await;
+
+        let response = match decrypted_payload {
+            Ok(payload) => DecryptionResponse::Ok(payload),
+            Err(err) => DecryptionResponse::Err(err),
+        };
+
+        // Send reply to the caller
+        ctx.send_from_address(return_route, response, self.addresses.decryptor_api.clone())
+            .await?;
+
+        Ok(())
+    }
+
+    pub(crate) async fn handle_decrypt(
+        &mut self,
+        ctx: &mut Context,
+        msg: Routed<Any>,
+    ) -> Result<()> {
+        debug!(
+            "SecureChannel {} received Decrypt {}",
+            self.role, &self.addresses.decryptor_remote
+        );
+
+        // Decode raw payload binary
+        let payload = Vec::<u8>::decode(&msg.into_transport_message().payload)?;
+
+        // Decrypt the binary
+        let decrypted_payload = self.decryptor.decrypt(&payload).await?;
+
+        // Encrypted data should be a TransportMessage
+        let mut transport_message = TransportMessage::decode(&decrypted_payload)?;
+
+        // Add encryptor hop in the return_route (instead of our address)
+        transport_message
+            .return_route
+            .modify()
+            .prepend(self.addresses.encryptor.clone());
+
+        // Mark message LocalInfo with IdentitySecureChannelLocalInfo,
+        // replacing any pre-existing entries
+        let local_info =
+            IdentitySecureChannelLocalInfo::mark(vec![], self.their_identity_id.clone())?;
+
+        let msg = LocalMessage::new(transport_message, local_info);
+
+        match ctx
+            .forward_from_address(msg, self.addresses.decryptor_internal.clone())
+            .await
+        {
+            Ok(_) => Ok(()),
+            Err(err) => {
+                warn!(
+                    "{} forwarding decrypted message from {}",
+                    err, &self.addresses.encryptor
+                );
+                Ok(())
+            }
+        }
+    }
+}
+
 pub(crate) struct Decryptor {
     current_key: KeyId,
     current_key_nonce: u64,
```

Comment:
I think you're just moving the code, so it's okay if you want to ignore my comments here, but I'll write my feedback here, we can move it elsewhere if needed

Reply:
The name `current_key_nonce` is misleading. It's not nonce at all, it can be called `number of rekys` or `epoch` or whatever

Reply:
Maybe it can be renamed to something like `last_rekey_nonce`

Reply:
Even better if we can could use `number_of_rekeys`, the math should be tweaked a little bit though